libslide/src/parser.rs

```diff
@@ -0,0 +1,135 @@
+use crate::scanner::types::Token;
+use crate::scanner::types::TokenType;
+
+pub enum Expr {
+    Float(f64),
+    Int(i64),
+    BinOp(BinOp),
+    // I added un op even though I havent implemented + or -
+    UnaryOp(UnaryOp),
+}
+
+pub struct BinOp {
+    pub item: Token,
+    pub lhs: Box<Expr>,
+    pub rhs: Box<Expr>,
+}
+
+pub struct UnaryOp{
+    pub item: Token, 
+    pub rhs: Box<Expr>,
+}
+
+pub struct Parser<'a>{
+    input: &'a Vec<Token>,
+    index: usize,
+    cur_token: Token,
+}
+
+impl <'a> Parser<'a>{
+    pub fn new (input: &Vec<Token>) -> Parser{
+        Parser{
+            input: input.into(),
+            index: 1,
+            cur_token: input[0].clone().into(),
+        }
+    }
+
+    fn get_token(&mut self) -> Token{
+        self.index = self.index+1;
+        return self.input[self.index-1].clone();
+    }
+
+    fn expr(&mut self) -> Box<Expr> {
+        return self.add_sub_term();
+    }
+
+    fn add_sub_term(&mut self) -> Box<Expr> {
+        let mut node = self.mul_divide_term();
+        if self.cur_token.token_type == TokenType::Plus || self.cur_token.token_type == TokenType::Minus{
+            let t = self.cur_token.clone();
+            self.cur_token = self.get_token();
+            node = Box::new(Expr::BinOp(BinOp{item: t, lhs: node, rhs: self.mul_divide_term()}));
+        }
+        return node;
+    }
+
+    fn mul_divide_term(&mut self) -> Box<Expr> {
+        let mut node = self.num_term();
+        if self.cur_token.token_type == TokenType::Mult || self.cur_token.token_type == TokenType::Div {
+            let t = self.cur_token.clone();
+            self.cur_token = self.get_token();
+            node = Box::new(Expr::BinOp(BinOp{item: t, lhs: node, rhs: self.num_term()}));
+        }
+        return node;
+    }
+
+    fn num_term(&mut self) -> Box<Expr> {
+        // 5.0 is a placeholder. mem::discriminant only compares variant types and ignores data
+        // this is pretty fucking cool rust has it
+        // this value should never be returned
+        let mut node = Box::new(Expr::Int(1));
+        if std::mem::discriminant(&self.cur_token.token_type) == std::mem::discriminant(&TokenType::Float(5.0)) { 
+            // look into a better way of doing this. only other method I can find is match
+            // by this I mean extract the data of variant float
+            if let TokenType::Float(f) = self.cur_token.token_type{
+                node = Box::new(Expr::Float(f));
+            }
```

Comment:
if you do pattern matching in the way i show in the comment above you can get this from the matched expression.